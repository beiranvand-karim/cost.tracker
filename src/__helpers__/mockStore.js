import configureStore from "redux-mock-store";
import costItems from "./costItems";

const mockStore = configureStore();

export default mockStore({
  costItems: {
    costItems: costItems
  },
  sourceCostItems: {
    costItems: costItems
  }
});
