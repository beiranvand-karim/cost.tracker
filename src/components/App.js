import React, { useState } from "react";
import { useSelector } from "react-redux";
import { Button } from "reactstrap";
import {
  ModalContainer,
  CostItemsTable,
  CostItemsTotalSum,
  CostItemCategoriesDropdown
} from ".";
import { Grid, Row } from "../styles";
import Shield from "./Shield";

export function App() {
  const [modal, setModal] = useState(false);
  const toggle = () => setModal(!modal);
  const costItems = useSelector(state => state.costItems.costItems);

  return (
    <Grid>
      <Row />
      <Row>
        <Button color="primary" onClick={toggle}>
          add a cost
        </Button>
      </Row>
      <Row>
        <Shield>
          <CostItemsTotalSum costItems={costItems} />
        </Shield>
      </Row>
      <Row>
        <Shield>
          <CostItemCategoriesDropdown />
        </Shield>
        <Shield>
          <ModalContainer toggle={toggle} modal={modal} />
        </Shield>
      </Row>
      <Row>
        <Shield>
          <CostItemsTable costItems={costItems} />
        </Shield>
      </Row>
    </Grid>
  );
}
