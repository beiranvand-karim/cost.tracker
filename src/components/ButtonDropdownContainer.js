import React from 'react';
import {categories} from '../config/categories';
import {useDispatch, useSelector} from 'react-redux';
import {addFilteredActions} from '../actions/filteredCostActions';
import NO_FILTER from "../constants/noFilter";

export default function ButtonDropdownButtonContainer() {

   const dispatch = useDispatch();
   const originalCost = useSelector(state => state.originalCosts);
   const {costs} = originalCost;

   const select = (event) => {
      const {value} = event.target;
      if (NO_FILTER === value) {
         dispatch(addFilteredActions(costs));
         return;
      }
      const filtered = costs.filter(cost => cost.category === value);
      dispatch(addFilteredActions(filtered));
   };

   const keys = Object.keys(categories);
   keys.unshift(NO_FILTER);

   return (
      <div>
         <select onChange={select}>
            {
               keys.map((category, index) => {
                  return <option key={index} value={category}>{category}</option>
               })
            }
         </select>
      </div>
   )
}
