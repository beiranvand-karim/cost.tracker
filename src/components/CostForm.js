import {Form, FormGroup, Input, Label} from 'reactstrap';
import {categories} from '../config/categories';
import React from 'react';

export default function CostForm({cost, setCost, category, setCategory, comment, setComment}) {
   return (
      <Form>
         <FormGroup>
            <Label for="exampleEmail">Cost</Label>
            <Input
               type="email"
               name="email"
               id="exampleEmail"
               placeholder="cost here..."
               value={cost}
               onChange={(e) => setCost(e.target.value)}
            />
         </FormGroup>
         <FormGroup>
            <Label for="exampleSelect">Select</Label>
            <Input
               type="select"
               name="select"
               id="exampleSelect"
               value={category}
               onChange={(e) => setCategory(e.target.value)}
            >
               {
                  Object.keys(categories).map((category, index) => <option key={index}>{category}</option>)
               }
            </Input>
         </FormGroup>
         <FormGroup>
            <Label for="examplePassword">Comment</Label>
            <Input
               type="input"
               name="comment"
               id="comment"
               value={comment}
               onChange={(e) => setComment(e.target.value)}
               placeholder="comment here..."
            />
         </FormGroup>
      </Form>
   )
}
