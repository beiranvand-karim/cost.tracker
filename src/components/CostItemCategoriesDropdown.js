import { useDispatch, useSelector } from "react-redux";
import React from "react";
import { NO_FILTER } from "../constants/constants";
import { addFilteredCostItemsAction } from "../actions";
import { costCategories } from "../config";

const costCategoriesKeys = () => {
  const temp = Object.keys(costCategories);
  temp.unshift(NO_FILTER);
  return temp;
};

export function CostItemCategoriesDropdown() {
  const dispatch = useDispatch();
  const sourceCostItems = useSelector(state => state.sourceCostItems.costItems);
  const handleChange = event => {
    const { value } = event.target;
    if (NO_FILTER === value) {
      dispatch(addFilteredCostItemsAction(sourceCostItems));
      return;
    }
    const CostItems = sourceCostItems.filter(
      costItem => costItem.category === value
    );
    dispatch(addFilteredCostItemsAction(CostItems));
  };

  return (
    <div>
      <select onChange={handleChange}>
        {costCategoriesKeys().map((costCategory, index) => {
          return (
            <option key={index} value={costCategory}>
              {costCategory}
            </option>
          );
        })}
      </select>
    </div>
  );
}
