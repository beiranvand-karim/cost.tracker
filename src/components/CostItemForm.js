import { Button, Form, FormGroup, Input, Label } from "reactstrap";
import { useDispatch } from "react-redux";
import React, { useState } from "react";
import { costCategories } from "../config";
import { addSourceCostItemAction, addCostItemAction } from "../actions";

const firstCategory = () => Object.keys(costCategories)[0];

export function CostItemForm({ toggle }) {
  const [cost, setCost] = useState("");
  const [comment, setComment] = useState("");
  const [category, setCategory] = useState(firstCategory());
  const dispatch = useDispatch();

  const handleSubmit = event => {
    event.preventDefault();
    setCost(cost);
    setComment(comment);
    const tempCostItem = { category, cost, comment };
    dispatch(addSourceCostItemAction(tempCostItem));
    dispatch(addCostItemAction(tempCostItem));
    clearForm();
    toggle();
  };

  const clearForm = () => {
    setCost("");
    setCategory(firstCategory());
    setComment("");
  };

  return (
    <Form onSubmit={handleSubmit}>
      <FormGroup>
        <Label for="cost">Cost</Label>
        <Input
          type="input"
          name="cost"
          id="cost"
          placeholder="cost here..."
          value={cost}
          onChange={event => setCost(event.target.value)}
        />
      </FormGroup>
      <FormGroup>
        <Label for="category">Category</Label>
        <Input
          type="select"
          name="category"
          id="category"
          value={category}
          onChange={event => setCategory(event.target.value)}
        >
          {Object.keys(costCategories).map((category, index) => (
            <option key={index}>{category}</option>
          ))}
        </Input>
      </FormGroup>
      <FormGroup>
        <Label for="comment">Comment</Label>
        <Input
          type="input"
          name="comment"
          id="comment"
          value={comment}
          onChange={event => setComment(event.target.value)}
          placeholder="comment here..."
        />
      </FormGroup>
      <Button color="primary" type="submit">
        add cost
      </Button>{" "}
      <Button color="secondary" onClick={toggle}>
        Cancel
      </Button>
    </Form>
  );
}
