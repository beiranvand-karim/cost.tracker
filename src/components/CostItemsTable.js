import { Table } from "reactstrap";
import React from "react";

export function CostItemsTable({ costItems }) {
  return (
    <Table responsive>
      <thead>
        <tr>
          <th>#</th>
          <th>Cost</th>
          <th>Category</th>
          <th>Comment</th>
        </tr>
      </thead>
      <tbody>
        {costItems.map((costItem, index) => (
          <tr key={index}>
            <th scope="row">{index}</th>
            <td>{costItem.cost}</td>
            <td>{costItem.category}</td>
            <td>{costItem.comment}</td>
          </tr>
        ))}
      </tbody>
    </Table>
  );
}
