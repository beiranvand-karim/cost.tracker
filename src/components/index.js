export { App } from "./App";
export { CostItemCategoriesDropdown } from "./CostItemCategoriesDropdown";
export { CostItemForm } from "./CostItemForm";
export { CostItemsTable } from "./CostItemsTable";
export { CostItemsTotalSum } from "./CostItemsTotalSum";
export { ModalContainer } from "./ModalContainer";
