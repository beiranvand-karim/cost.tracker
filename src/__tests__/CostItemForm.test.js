/* global expect, describe, it, beforeEach */
import { costCategories } from "../config";
import { CostItemForm } from "../components/CostItemForm";
import mockStore from "../__helpers__/mockStore";
import costItems from "../__helpers__/costItems";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";
import { Provider } from "react-redux";
import React from "react";

describe("<CostItemForm/>", () => {
  const cost = costItems[0].cost;
  const comment = costItems[0].comment;
  const category = costItems[0].category;
  const toggleMock = jest.fn();
  const store = mockStore;

  beforeEach(() => {
    store.clearActions();
  });

  it("should render without crashing", () => {
    shallow(
      <Provider store={store}>
        <CostItemForm />
      </Provider>
    );
  });

  it("should add a cost item", () => {
    const costForm = mount(
      <Provider store={store}>
        <CostItemForm toggle={toggleMock} />
      </Provider>
    );
    costForm
      .find("#cost")
      .at(1)
      .simulate("change", { target: { value: cost, name: "cost" } });
    costForm
      .find("#comment")
      .at(1)
      .simulate("change", { target: { value: comment, name: "comment" } });
    costForm
      .find("#category")
      .at(1)
      .simulate("change", { target: { value: category, name: "category" } });
    costForm.find("Form").simulate("submit", { preventDefault() {} });
    expect(store.getActions()).toMatchSnapshot();
    expect(toggleMock).toHaveBeenCalled();
  });

  it("should render cost categories in a select tag", () => {
    const wrapper = mount(
      <Provider store={store}>
        <CostItemForm cost={cost} comment={comment} category={category} />
      </Provider>
    );

    for (let i = 0; i < Object.keys(costCategories).length; i++) {
      expect(
        wrapper
          .find("option")
          .at(i)
          .text()
      ).toBe(Object.keys(costCategories)[i]);
    }
  });

  it("should matches the snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <CostItemForm cost={cost} comment={comment} category={category} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
