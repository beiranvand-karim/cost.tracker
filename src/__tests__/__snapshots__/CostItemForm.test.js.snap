// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CostItemForm/> should add a cost item 1`] = `
Array [
  Object {
    "payload": Object {
      "category": "CLOTHING",
      "comment": "this is a comment",
      "cost": 1000,
    },
    "type": "ADD_SOURCE_COST_ITEM",
  },
  Object {
    "payload": Object {
      "category": "CLOTHING",
      "comment": "this is a comment",
      "cost": 1000,
    },
    "type": "ADD_COST_ITEM",
  },
]
`;

exports[`<CostItemForm/> should matches the snapshot 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="cost"
    >
      Cost
    </label>
    <input
      className="form-control"
      id="cost"
      name="cost"
      onChange={[Function]}
      placeholder="cost here..."
      type="input"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="category"
    >
      Category
    </label>
    <select
      className="form-control"
      id="category"
      name="category"
      onChange={[Function]}
      value="FOOD"
    >
      <option>
        FOOD
      </option>
      <option>
        CLOTHING
      </option>
      <option>
        RENT
      </option>
      <option>
        BILLS
      </option>
    </select>
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="comment"
    >
      Comment
    </label>
    <input
      className="form-control"
      id="comment"
      name="comment"
      onChange={[Function]}
      placeholder="comment here..."
      type="input"
      value=""
    />
  </div>
  <button
    aria-label={null}
    className="btn btn-primary"
    onClick={[Function]}
    type="submit"
  >
    add cost
  </button>
   
  <button
    aria-label={null}
    className="btn btn-secondary"
    onClick={[Function]}
  >
    Cancel
  </button>
</form>
`;
